tests/integration/helpers/cluster.py

```diff
@@ -1259,9 +1261,22 @@ def wait_for_start(self, deadline=None, timeout=None):
                     "Instance `{}' failed to start. Container status: {}, logs: {}".format(self.name, status,
                                                                                            handle.logs().decode('utf-8')))
 
+            current_deadline = deadline
+            # It is possible that server starts slowly.
+            # If container is running, and there is some progress in log, check connection_timeout.
+            if connection_deadline and current_deadline < connection_deadline and status == 'running':
+                rows_in_log = self.count_in_log(".*").strip()
+                try:
+                    rows_in_log = int(rows_in_log)
+                    if rows_in_log > prev_rows_in_log:
+                        prev_rows_in_log = rows_in_log
+                        current_deadline = connection_deadline
```

Comment:
Strange, but it looks like this `if` is always `true` on the first iteration (`whatever > 0`) and does nothing after the first iteration (`current_deadline` already equal to `connection_deadline`).

Reply:
> it looks like this if is always true on the first iteration

It is not true if logs are empty.

> `current_deadline` already equal to `connection_deadline`

But we set `current_deadline = deadline` in line 1264 ...

Reply:
However, it is not good that we wait for too long

Reply:
Now this code is unreadable :(